### DISCUSSION TITLE
How to force nuget packages to be restored when installing my vcpkg package?

### DISCUSSION THREAD

**Discussion by lukcz** [microsoft/vcpkg#21200] [Q&A]
I am preparing a custom port for my own library, but it fails during installation. In the log file, I can see that it is missing NuGet packages which it has dependencies to. How I can force them to be restored during installation, when calling:
```
vcpkg install custom_package
```
??

Error message:
```
error : This project references NuGet package(s) that are missing on this computer. Use NuGet Package Restore to download them.  For more information, see http://go.microsoft.com/fwlink/?LinkID=322105. The missing file is <missing_file>
Done Building Project "XXX.vcxproj" (Rebuild target(s)) -- FAILED.

Build FAILED.
```

In script vcpkg_install_msbuild.cmake I can see that build is called with "/t:Rebuild". In order to make it work it would have to be called with "/t:Restore;Rebuild" parameter, right?

```
 list(APPEND arg_OPTIONS
        "/t:Rebuild"
        "/p:Platform=${arg_PLATFORM}"
        "/p:PlatformToolset=${arg_PLATFORM_TOOLSET}"
        "/p:VCPkgLocalAppDataDisabled=true"
        "/p:UseIntelMKL=No"
        "/p:WindowsTargetPlatformVersion=${arg_TARGET_PLATFORM_VERSION}"
        "/p:VcpkgTriplet=${TARGET_TRIPLET}"
        "/p:VcpkgInstalledDir=${_VCPKG_INSTALLED_DIR}"
        "/p:VcpkgManifestInstall=false"
        "/m"
    )
```
After changing it I can see that build is being called with:
```
"/t:Restore Rebuild "
```
instead of (semicolon is being removed somewhere along the way in vcpkg_execute_required_process.cmake)
``` 
"/t:Restore;Rebuild" 
```

Any ideas, how it can be solved?

**Comment by lukcz** [comment]
So I found a workaround by adding:
```
        "/r:True"
	"/p:RestorePackagesConfig=True"
```

into arg_OPTIONS:

```
list(APPEND arg_OPTIONS
        "/t:${arg_TARGET}"
        "/r:True"
	"/p:RestorePackagesConfig=True"
	"/p:Platform=${arg_PLATFORM}"
        "/p:PlatformToolset=${arg_PLATFORM_TOOLSET}"
        "/p:VCPkgLocalAppDataDisabled=true"
        "/p:UseIntelMKL=No"
        "/p:WindowsTargetPlatformVersion=${arg_TARGET_PLATFORM_VERSION}"
        "/p:VcpkgTriplet=${TARGET_TRIPLET}"
        "/p:VcpkgInstalledDir=${_VCPKG_INSTALLED_DIR}"
        "/p:VcpkgManifestInstall=false"
        "/m"
    )
```
and I am wondering if it shouldn't be enabled by default?

**Comment by NancyLi1013** [comment]
Hi @lukcz 

Thanks for posting this issue.
Since you used a custom package to build, the package depends on a NuGet package. Currently, vcpkg only supports the dependencies in these situations:

- The dependent packages are provided by vcpkg.
- The dependent packages are embed in the source codes.
- The dependent packages have been downloaded to your local machine.

For your cases, you can use above workaround to solve your problems or download the NuGet package to your local. If the port has been added to vcpkg, you can also use it.

**Comment by lukcz** [comment]
Hi,
well restoring the custom package is one thing, but the second (even more important for me) is the integration of vcpkg with VS Project using the plugin (NuGet package) which is generated after calling:
```
vcpkg integrate project
```
In fact this plugin is just NuGet dependency which can be added into the project. Having prior defined additional NuGet package source (pointing to location where command "vcpkg integrate project" has written its output) this plugin could be easily restorred when project compilation starts.

This approach was described here (in section "Linking NuGet file"):
https://github.com/microsoft/vcpkg/blob/master/docs/users/integration.md#with-msbuild

The error I mentioned at the beginning  shows up when in following scenario.

I have two projects:

Project A (uses vcpkg in manifest mode, defined vcpkg.json)
Project B (uses vcpkg in manifest mode, defined  vcpkg.json)

Project A, requires Project B as its dependency (declared in Project A's vcpkg.json).

When I am compiling Project A, then vcpkg tries to resolve dependency to Project B, however it fails.
It fails due to problem with obtaining NuGet package (needed for vcpkg integration) when Project B is being compiled.

Project B in its portile.cmake is calling vcpkg_install_msbuild.cmake(), but as I already mentioned vcpkg_install_msbuild scirpt is missing those two parameters to restore NuGet dependency:

```
        "/r:True"
	"/p:RestorePackagesConfig=True"
```

Then NuGet Dependency is not restored for project B, and whole compilation fails.

**Comment by NancyLi1013** [comment]
@strega-nil Do you have any ideas about this issue?
